algo-core/scheduler_utils.py

```diff
@@ -0,0 +1,156 @@
+
+def tracks_hours_to_slots(tracks):
+    for track in tracks:
+        track["start_slot"] = track["start_hour"] * 2
+        track["end_slot"] = track["end_hour"] * 2
+    return tracks
+
+def slots_to_range(week_hours, end_hour):
+    """Convert per-hour availability flags into ranges format."""
+    week_ranges = []
+
+    for day_hours in week_hours:
+        day_ranges = []
+        start = None
+
+        for i, value in enumerate(day_hours):
+            # Start of new range:
+            if start is None and value != 0:
+                start = i
+                continue
+
+            # End of range:
+            # (A range will end if either the current slot is unavailable
+            # (value 0) or if the current slot is the last one.)
+            if start is not None:
+                if value == 0:  # Unavailable
+                    day_ranges.append([start, i])
+                    start = None
+                elif i == end_hour - 1:  # Last slot
+                    day_ranges.append([start, end_hour])
+                else:
+                    continue
+
+        week_ranges.append(day_ranges)
+
+    return week_ranges
+
+
+def print_final_schedules(schedule_results, df_agents, num_days):
+    """Print final schedule, validate output JSON, and write to file."""
+    for d in range(num_days):
+        print(
+            f"\n{schedule_results[d]['start_date'].strftime('%Y-%m-%d')} "
+            "shifts:"
+        )
+
+        for (i, e) in enumerate(schedule_results[d]["shifts"]):
+            print(e)
+
+    output_json = []
+
+    for epoch in schedule_results:
+        # Substitute agent info from 'handle' to 'handle <email>'
+        shifts = []
+
+        for (name, start, end) in epoch["shifts"]:
+            shifts.append(
+                {
+                    "agent": f"{name} <{df_agents.loc[name, 'email']}>",
+                    "start": start,
+                    "end": end,
+                }
+            )
+
+        day_dict = {"start_date": epoch["start_date"].strftime("%Y-%m-%d"), "shifts": shifts}
+        output_json.append(day_dict)
+
+    output_json_schema = json.load(
+        open(get_project_root() + "/lib/schemas/support-shift-scheduler-output.schema.json")
+        )
+
+    try:
+        jsonschema.validate(output_json, output_json_schema)
+    except jsonschema.exceptions.ValidationError as err:
+        print("Output JSON validation error", err)
+        sys.exit(1)
+
+    print("\nSuccessfully validated JSON output.")
+
+    with open(input_folder + "support-shift-scheduler-output.json", "w") as outfile:
+        outfile.write(json.dumps(output_json, indent=4))
+
+    return output_json
+
+
+def read_onboarding_files():
+    """Read agent handles from onboarding-related files into pandas series."""
+
+    if Path(input_folder + filename_onboarding).exists():
+        ser_o = pd.read_csv(
+            input_folder + filename_onboarding, squeeze=True, header=None, names=["agents"]
+        )
```

Comment:
In all of the cases it's possible to define the path once and check for its existence as well as pass it to `Pandas.read_csv` function.

Using python 3.8+, the [walrus operator](https://www.python.org/dev/peps/pep-0572/#examples) seems like a good fit here:
```suggestion
    if (path := Path(input_folder, filename_onboarding)).exists():
        ser_o = pd.read_csv(path, squeeze=True, header=None, names=["agents"])
```

Reply:
nice!! didn't know that